### PR TITLE
Programming_Engine: Adding Initial code

### PR DESCRIPTION
**Pull request by adecler**
### NOTE: Depends on 
https://github.com/BHoM/BHoM/pull/682

   
Provides the Engine for the task covered by this PR: https://github.com/BHoM/Grasshopper_Toolkit/pull/447

This effectively removes the dependency of Grasshopper with the Node2Code toolkit. It is critical to do so since some of the code in GH related to Node2Code will later be migrated to BHoM_UI.

In term of review, this is mainly a question of testing that the Grasshopper PR still works the same ways now that the oM has been migrated.

### Additional comments
This is simply a migration of the code from https://github.com/BHoM/Node2Code_Toolkit/pull/2 with no change other than the namespace

### CODE REVIEW

**Comment by adecler** [comment]
@FraserGreenroyd , AppVeyor is happy but not DevOps. I suppose this is the same problem as for https://github.com/BHoM/BHoM_Engine/pull/1449

File: Programming_Engine/Compute/PopulateTypes.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+            if (node.Outputs.Count > 0)
+                node.Outputs[0].DataType = node.ObjectType;
+
+            return node;
+        }
+
+        /***************************************************/
+
+        public static TypeNode PopulateTypes(this TypeNode node)
```
Comment (FraserGreenroyd) [inline Programming_Engine/Compute/PopulateTypes.cs]: Description, input, and output attributes needed here

File: Programming_Engine/Compute/PopulateTypes.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+                param.DataType = typeof(Type);
+
+            return node;
+        }
+
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        public static INode PopulateTypes(this INode node)
```
Comment (FraserGreenroyd) [inline Programming_Engine/Compute/PopulateTypes.cs]: If this is meant to be public then Description, input, and output attributes needed here.

However, it is under the heading "Private Methods" - so I guess I'm querying whether it should be private instead? If it is private then obviously the above attributes aren't needed

File: Programming_Engine/Programming_Engine.csproj [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="BHoM, Version=3.0.0.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (FraserGreenroyd) [inline Programming_Engine/Programming_Engine.csproj]: Reference shouldn't include version, culture, or processor architecture

File: Programming_Engine/Programming_Engine.csproj [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>..\..\BHoM\Build\BHoM.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
+    <Reference Include="Programming_oM">
+      <HintPath>..\..\BHoM\Build\Programming_oM.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
+    <Reference Include="Reflection_oM, Version=3.0.0.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (FraserGreenroyd) [inline Programming_Engine/Programming_Engine.csproj]: Reference shouldn't include version, culture, or processor architecture

File: Programming_Engine/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
```
Comment (FraserGreenroyd) [inline Programming_Engine/Properties/AssemblyInfo.cs]: Copyright header needed (as it is a CS file and consistent with our other approaches)

File: Programming_Engine/Query/DataTypePerParam.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
+﻿using BH.oM.Programming;
```
Comment (FraserGreenroyd) [inline Programming_Engine/Query/DataTypePerParam.cs]: Copyright header needed

File: Programming_Engine/Query/DataTypePerParam.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace BH.Engine.Programming
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Interface Methods                         ****/
+        /***************************************************/
+
+        public static Dictionary<Guid, Type> IDataTypePerParam(this INode node)
```
Comment (FraserGreenroyd) [inline Programming_Engine/Query/DataTypePerParam.cs]: Description, input, and output attributes needed here

File: Programming_Engine/Query/DataTypePerParam.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            return DataTypePerParam(node as dynamic);
+        }
+
+
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static Dictionary<Guid, Type> DataTypePerParam(this ClusterContent content)
```
Comment (FraserGreenroyd) [inline Programming_Engine/Query/DataTypePerParam.cs]: Description, input, and output attributes needed here

File: Programming_Engine/Query/DataTypePerParam.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                foreach (KeyValuePair<Guid, Type> kvp in childTypes)
+                    types[kvp.Key] = kvp.Value;
+            }
+
+            return types;
+        }
+
+        /***************************************************/
+
+        public static Dictionary<Guid, Type> DataTypePerParam(this INode node)
```
Comment (FraserGreenroyd) [inline Programming_Engine/Query/DataTypePerParam.cs]: Description, input, and output attributes needed here

**Comment by adecler** [comment]
Sorry about that. You're definitely right @FraserGreenroyd , this should be taken core in this PR. 
Now corrected.

**Comment by adecler** [comment]
@FraserGreenroyd , It looks like there is still something odd with CI as it seems to use the oM from master

**Comment by FraserGreenroyd** [comment]
> @FraserGreenroyd , It looks like there is still something odd with CI as it seems to use the oM from master

It appears to be trying to find the link to the new quantities stuff that @IsakNaslundBh added yesterday 🤔 just trying to track down why it thinks it wants to do that in the PS files now

**Comment by al-fisher** [comment]
Hey @adecler - unless I have missed something - I think you need to migrate some more methods across to `BH.Engine.Programming` here from the Node2Code repo ?

Can not find new counter part to the old `BH.Engine.Node2Code.Convert.ToCSharpText` - taking updated `BH.oM.Programming.ClusterContent`
image

**Comment by FraserGreenroyd** [comment]
@adecler the CI should now be rectified - I have a green tick at the moment, so any future red crosses should probably be double checked in case of an issue with the PR :smile: sorry for the delay/confusion!

**Comment by adecler** [comment]
Hey @al-fisher, the idea is that the creation of the BHoM syntax tree is now in the core BHoM. Everything that then converts this syntax tree into another language (C#, GH & Dynamo script, ...) will be in a separate toolkit. The conversion to C# (i.e. what is left of the node2code toolkit) will later be migrated to the C# toolkit. I just didn't do it yet because it felt odd merging the code in there with the node2code stuff as they serve very different purposes. So wanted to discuss first.

**Comment by al-fisher** [comment]
Right - sorry - yep. Makes sense.

File: Programming_Engine/Compute/PopulateTypes.cs [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+{
+    public static partial class Compute
+    {
+        /***************************************************/
+        /**** Interface Methods                         ****/
+        /***************************************************/
+
+        [Description("Automatically set the DataType and default value of all parameters of the input syntax node.")]
+        [Input("node", "syntax node to modify")]
+        [Output("Modified node")]
```
Comment (FraserGreenroyd) [inline Programming_Engine/Compute/PopulateTypes.cs]: If the output is a modified node, would this not be a modify class method rather than compute?

Comment (al-fisher) [reply Programming_Engine/Compute/PopulateTypes.cs]: Yep - quite possibly. 
@adecler and I picking up a couple of other code and structural organisation for node2code and soon to be public CSharp toolkit - which I think could relate a little to best structuring of Programming_oM. Maybe. Happy to merge and mop up later if needed

Comment (FraserGreenroyd) [reply Programming_Engine/Compute/PopulateTypes.cs]: SGTM, am happy with that approach :smile:

**Pull request merged**